modules/nf-commons/src/main/nextflow/util/HashBuilder.java

```diff
@@ -324,41 +324,47 @@ static protected Hasher hashFileSha256( Hasher hasher, Path path, Path base ) {
 
     static protected Hasher hashDirSha256( Hasher hasher, Path dir, Path base ) {
         try {
+            var entries = new HashMap<String, String>();
+
             Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+                @Override
                 public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
                     log.trace("Hash sha-256 dir content [FILE] path={} - base={}", path, base);
                     try {
-                        // the file relative base
-                        if( base!=null )
-                            hasher.putUnencodedChars(base.relativize(path).toString());
+                        // the file relative path
+                        var name = relativePath(path, base).toString();
                         // the file content sha-256 checksum
-                        String sha256 = sha256Cache.get(path);
-                        hasher.putUnencodedChars(sha256);
+                        var value = sha256Cache.get(path);
+                        entries.put(name, value);
                         return FileVisitResult.CONTINUE;
                     }
                     catch (ExecutionException t) {
-                        throw new IOException(t);
+                        throw new IOException("Failed to get SHA-256 from cache for " + path, t);
                     }
                 }
 
+                @Override
                 public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes attrs) {
                     log.trace("Hash sha-256 dir content [DIR] path={} - base={}", path, base);
-                    // the file relative base
-                    if( base!=null )
-                        hasher.putUnencodedChars(base.relativize(path).toString());
-                    hasher.putUnencodedChars(base.relativize(path).toString());
+                    entries.put(relativePath(path, base).toString(), null);
                     return FileVisitResult.CONTINUE;
                 }
             });
+
+            return hashUnorderedCollection(hasher, entries.entrySet(), HashMode.SHA256);
```

Comment:
That's a good idea because bring down the complexity from O(n*log(n)) to 2*O(n). However it could be further optimised implementing the `hashUnorderedCollection` (which is quite simple) in the `walkFileTree` block (also the HashMap would not needed anymore). This could avoid avoid requiring a separate loop to compute the hash.

Reply:
Good idea. I found a way to eliminate the intermediate hashmap and still re-use `hashUnorderedCollection()`